File: Python/euler83.py
import numpy as np

'''
	Read File into a 2D Matrix
'''
filepath='C:\\Users\\kiit1\\Documents\\Codes\\Project-Euler\\euler83.txt'
matrix_ = list()
with open(filepath, 'r') as filehandle:  
	for line in filehandle:  
		currentPlace = line.split('\n')
		res=list(currentPlace[0].split(','))
		final_res=list(map(int,res))
		matrix_.append(final_res)


matrix=np.matrix(matrix_)

''' Global Variables '''

ROW,COL=matrix.shape[0],matrix.shape[1]
visited=np.matrix([[False]*COL]*ROW)	
row_queue,col_queue=[],[]
row_,col_=0,0
reached_dest=False
#minimal_sum=matrix[row_,col_]
#move_count=0
#nodes_left_in_layer, nodes_in_next_layer=1,0

direction_row=[-1,+1,0,0]
direction_col=[0,0,+1,-1]
	

def find_minimal_path_using_bfs():	
	nodes_left_in_layer, nodes_in_next_layer=1,0
	move_count=0
	row_queue.append(row_)
	col_queue.append(col_)
	visited[row_,col_]=True
	while len(row_queue)>0:
		r=row_queue.pop(0)
		c=col_queue.pop(0)
		if matrix[r,c]==matrix[ROW-1,COL-1]:
			reached_dest=True
			break
		explore_neighbours(r,c)
		nodes_left_in_layer-=1
		if nodes_left_in_layer==0:
			nodes_left_in_layer=nodes_in_next_layer
			nodes_in_next_layer=0
			move_count+=1
	if reached_dest:
		return move_count
	return-1

def explore_neighbours(r,c):
	nodes_in_next_layer=0
	for x in range(0,4):
		rr=r+direction_row[x]
		cc=c+direction_col[x]

		#skipping out of bound location
		if rr<0 or cc<0:
			continue
		if rr>=ROW or cc>=COL:
			continue
		#Skipping visited location
		if visited[rr,cc]:
			continue

		row_queue.append(rr)
		col_queue.append(cc)
		visited[rr,cc]=True
		#print(matrix[rr,cc])
		nodes_in_next_layer+=1


def solve():
	count_=find_minimal_path_using_bfs()
	print(count_)

if __name__ == '__main__':
	solve()

File: C#/euler83.cs
using System;
using System.Collections.Generic;
using System.Collections;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace ProjectEuler
{
    class euler83
    {
        const string filepath = @"C:\Users\kiit1\Documents\Codes\Project-Euler\euler83.txt";
        const int ROWS = 80;
        const int COLS = 80;
        static readonly int[] direction_rows= { +1, -1, 0, 0};
        static readonly int[] direction_cols = { 0, 0, +1, -1 };

        private static int[,] readFile(string filepath)
        {
            int[,] adj_matrix = new int[ROWS,COLS];
            int row_ = 0, col_ = 0;
            using (StreamReader reader = File.OpenText(filepath))
            {
                string line;
                while ((line = reader.ReadLine()) != null)
                {
                    string[] row = line.Split(',');
                    foreach (string col in row)
                    {
                        adj_matrix[row_, col_] = int.Parse(col);
                        col_++;
                    }
                    row_++;
                    col_ = 0;
                }
            }

            return adj_matrix;
        }

        private static int solve(int[,] adjMatrix)
        {
            Tuple<int,int> current_tup = new Tuple<int,int>(0, 0);
            Dictionary<Tuple<int, int>, int> permanent = new Dictionary<Tuple<int, int>, int>();

            permanent.Add(current_tup,weight(current_tup));

            Dictionary<Tuple<int, int>, int> temporary = new Dictionary<Tuple<int, int>, int>();
            int total_nodes = ROWS * COLS;

            while ((permanent.Count)< total_nodes)
            {
                List<Tuple<int, int>> neighbours = find_neighbours(current_tup);
                foreach (Tuple<int,int> node in neighbours)
                {
                    
                    if (!permanent.ContainsKey(node))
                    {
                        if (temporary.ContainsKey(node))
                        {
                            if (temporary[node] > permanent[current_tup] + weight(node))
                            {
                                temporary[node] = permanent[current_tup] + weight(node);
                            }
                        }
                        else
                        {
                            try
                            {
                                temporary.Add(node, permanent[current_tup] + weight(node));
                            }
                            catch (Exception e)
                            {
                                Console.WriteLine(e.Message);
                            }                            
                        }
                    }                    
                }

                int min_weight = temporary.Values.Min();
                List<Tuple<int, int>> min_nodes = new List<Tuple<int, int>>();
                foreach (KeyValuePair<Tuple<int,int>,int> key_val in temporary)
                {
                    if (key_val.Value == min_weight)
                    {
                        min_nodes.Add(key_val.Key);
                    }
                }

                Tuple<int, int> min_singular_node = min_nodes[0];
                permanent[min_singular_node] = min_weight;
                temporary.Remove(min_singular_node);
                current_tup = min_singular_node;

            }
            return permanent[new Tuple<int,int>(ROWS-1,COLS-1)];
        }

        private static List<Tuple<int,int>> find_neighbours(Tuple<int,int> tup)
        {
            var indices = tup;
            int r = indices.Item1;
            int c = indices.Item2;
            List<Tuple<int, int>> moves = new List<Tuple<int, int>>();
            for (int x = 0; x < 4; x++)
            {
                int rr = r + direction_rows[x];
                int cc = c + direction_cols[x];

                if (rr<0 || cc < 0)
                {
                    continue;
                }
                if (rr>=ROWS || cc >= COLS)
                {
                    continue;
                }
                moves.Add(Tuple.Create(rr, cc));
            }
            return moves;
        }

        private static int weight(Tuple<int, int> tup)
        {
            int row = tup.Item1;
            int col = tup.Item2;

            int[,] matrix = readFile(filepath);
            return matrix[row,col];
        }


        public static void Main(string[] args)
        {            
            int[,] adj_mat = readFile(filepath);
            int result = solve(adj_mat);
            Console.WriteLine(result);
            Console.ReadKey();
        }
    }
}
